### PR TITLE
Make project descriptions editable in the DiscoverySidebar

### PR DESCRIPTION
**Pull request by j-x-han**
# Description

Users can now edit the descriptions of projects they have access to by clicking the grey "Edit" button in the DiscoverySidebar after a project has been selected.

image

The component follows the pattern of the editable MetadataSections in the sample details sidebar.

image

Users will not be able to edit the project descriptions of any project (public or private) that they have not been added to. The grey "Edit" button will not appear in these cases

Without access to a public project:

image

Without access to a private project:

image

# Tests

* Verified that users cannot edit project descriptions to be over 700 characters.
* Verified that HTML tags are removed from project descriptions.
* Created non-admin test account and verified that it cannot edit project descriptions without project access
    * On the front-end: the "Edit" button will not be displayed
    * On the back-end: the "update" action will be ignored if the current user does not have update access to the project

### CODE REVIEW

File: app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx [hunk 1]
```diff
@@ -3,26 +3,81 @@ import cx from "classnames";
...
+  _save = async (projectId, value) => {
+    this.setState({
+      savePending: true,
+    });
+
+    let lastValidDescription = this.state.lastValidDescription;
+
+    const response = await saveProjectDescription(projectId, value);
+
+    // If save fails, revert to last valid description value.
```
Comment (gregdingle) [inline app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx]: why not show an error? and why only handle 422 errors?

File: app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx [hunk 2]
```diff
@@ -31,34 +86,56 @@ class ProjectDescription extends React.Component {
       : false;
 
     return (
-      <Accordion
-        className={cs.descriptionContainer}
-        bottomContentPadding
+      <MetadataSection
+        editable={this.props.project.editable}
+        onEditToggle={() => this.toggleEditing()}
```
Comment (gregdingle) [inline app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx]: shorten to `onEditToggle=[this.toggleEditing}` ?

File: app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx [hunk 3]
```diff
@@ -31,34 +86,56 @@ class ProjectDescription extends React.Component {
...
-            <div
-              className={cx(
-                shouldTruncateDescription && showLess && cs.truncated
-              )}
-            >
-              {description}
+        {this.state.editing ? (
+          <div className={cs.descriptionContainer}>
+            <Textarea
+              onChange={val => this.handleDescriptionChange(val)}
```
Comment (gregdingle) [inline app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx]: shorten these handler expressions as above

File: app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx [hunk 4]
```diff
@@ -3,26 +3,81 @@ import cx from "classnames";
...
+          changed: false,
+        },
+        () => {
+          this._save(this.props.project.id, this.state.description);
+        }
+      );
+    }
+  };
+
+  _save = async (projectId, value) => {
```
Comment (gregdingle) [inline app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx]: I don't think we follow the underscore convention for private JS methods, though I like it. I'd like to hear others's thoughts.

Comment (j-x-han) [reply app/assets/src/components/views/discovery/DiscoverySidebar/ProjectDescription.jsx]: It's used in `SampleDetailsMode.jsx` to save metadata details as well, which is the style I was following!

**Pull request merged**